pandas/core/groupby/groupby.py

```diff
@@ -764,7 +766,19 @@ def _try_cast(self, result, obj, numeric_only=False):
             dtype = obj.dtype
 
         if not is_scalar(result):
-            if is_extension_array_dtype(dtype):
+            if is_datetime64tz_dtype(dtype):
+                # GH 23683
+                # Prior results _may_ have been generated in UTC.
+                # Ensure we localize to UTC first before converting
+                # to the target timezone
+                try:
+                    utc_dtype = DatetimeTZDtype(tz='UTC')
+                    result = obj._values._from_sequence(result,
+                                                        dtype=utc_dtype)
+                    result = result.astype(dtype)
+                except TypeError:
```

Comment:
when does this hit the type error? can you add a comment about what types of returns yield this (assume it is when the transform function is doing a non-time transform). is this hit in the tests?

Reply:
Added a comment when this was hit (`_try_cast` is called downstream where the result is already correctly localized). Tested here:

https://github.com/pandas-dev/pandas/blob/b2c751985ebd09b72d917d08dc06193dc0922018/pandas/tests/groupby/test_transform.py#L840